Initial `/search/transactions` (Currently Only `user_command`)

piotr-iohk opened the pull request:
Gives roughly the same response between Rosetta (`https://rosetta-online-mainnet.minaprotocol.network/search/transactions`) and MinaMesh (`http://localhost:3000/search/transactions`) for something like:

```
$ curl -L -X POST '<endpoint here>'   \    
 -H 'Content-Type: application/json'   \  
 -H 'Accept: application/json' -d '{
  "network_identifier": {
                "blockchain": "mina",
                "network": "mainnet"
  },
  "address": "B62qkHazss1uxZ8iFFxkfTjCPPg1ApXHax3w5FQzQcqpaR2hQRBi3uV", "limit":25
}'
```

harrysolovay commented on src/api/search_transactions.rs:
```diff
@@ -0,0 +1,254 @@
+    operations.push(operation(
+      operation_index,
+      Some(&format!("-{}", self.fee.unwrap_or_else(|| "0".to_string()))),
+      AccountIdentifier {
```
Shall we use `AccountIdentifier::new` for consistency?

piotr-iohk replied:
Unfortunately `AccountIdentifier::new` only takes address as a parameter and we are also having metadata here.

harrysolovay commented on src/api/search_transactions.rs:
```diff
@@ -0,0 +1,254 @@
+
+    // Operation 2: Account Creation Fee (if applicable)
+    if let Some(creation_fee) = &self.creation_fee {
+      if let Ok(fee_value) = creation_fee.parse::<i64>() {
```
Should we map the error into a `MinaMeshError` and then try/`?` it (instead of not handling the parse error case)?

piotr-iohk replied:
Actually we don't need to parse/cast it indeed. The `creation_fee` comes from Archive DB where it is hold as text, so better to present it as is. Corrected in 739e291282337ee4af26adf41304637dc802ea9e

harrysolovay commented on src/api/search_transactions.rs:
```diff
@@ -0,0 +1,254 @@
+          operations.push(operation(
+            operation_index,
+            Some(&format!("-{}", creation_fee)),
+            AccountIdentifier {
```
Same comment re. `AccountIdentifier::new`. I see a few more occurrences below as well.

piotr-iohk replied:
Same: https://github.com/MinaFoundation/MinaMesh/pull/32#discussion_r1791257989

harrysolovay commented on src/api/search_transactions.rs:
```diff
@@ -0,0 +1,254 @@
+    // Decide on the type of operation based on command type
+    match self.command_type {
+      UserCommandType::Payment => {
+        // Operation 3: Payment Source Decrement
```
Stylistic preference: can we place comments above the match arm?

piotr-iohk replied:
Done in 739e291282337ee4af26adf41304637dc802ea9e.

harrysolovay commented on src/api/search_transactions.rs:
```diff
@@ -0,0 +1,254 @@
+      offset,
+    )
+    .fetch_all(&self.pg_pool)
+    .await;
```
We've implemented `From<SqlxError>` for `MinaMeshCommand`, so you should be able to just try/`?` the result (instead of needing to match below).

piotr-iohk replied:
Done in 650b98833d7d18077300e8f54c5de85b88aca6e4.

harrysolovay commented on src/api/search_transactions.rs:
```diff
@@ -0,0 +1,254 @@
+  }
+
+  #[allow(dead_code)]
+  async fn fetch_internal_commands(
```
What's the game plan re. these two fns?

piotr-iohk replied:
The plan is to implement them and concatenate results with `fetch_user_commands` result which would be the ultimate response for `search_transactions`.

harrysolovay commented on src/operation.rs:
```diff
@@ -0,0 +1,28 @@
+
+use crate::{OperationStatus, OperationType, TransactionStatus};
+
+pub fn operation(
```
In other parts of the code, I see quite a bit of `clone`ing of arguments supplied to this factory fn. Should we instead update the parameters to be references? This way we could get rid of some of the boilerplate elsewhere.

harrysolovay reviewed (changes requested):
@piotr-iohk had a few small comments. However, it's looking really, really good.

piotr-iohk commented:
@harrysolovay I've addressed the review comments. Also created https://github.com/MinaFoundation/MinaMesh/issues/48 to document the finding we discussed the other day.

harrysolovay commented on src/error.rs:
```diff
@@ -7,6 +7,9 @@ use thiserror::Error;
 #[derive(Error, Debug, PartialEq)]
 pub enum MinaMeshError {
+  #[error("Invalid transaction status: '{0}'. Valid are 'applied' and 'failed'")]
+  InvalidTransactionStatus(String),
```
@joaosreis is it okay for us to introduce new error variants that weren't in the initial Rosetta implementation?

piotr-iohk replied:
Just to note in Rosetta, when using invalid `status` in `/search/transactions` it throws SQL failure, which doesn't seem nice:
```
{
  "code": 1,
  "message": "SQL failure",
  "description": "We encountered a SQL failure.",
  "retriable": false,
  "details": {
    "body": [
      "Sql",
      "Request to <postgres://mina:_@archive-postgresql:5432/archive> failed: ERROR:  invalid input value for enum transaction_status:
```

harrysolovay reviewed (approved):
Awesome work!
